### PR TITLE
8288050: Add support of SHA-512/224 and SHA-512/256 to the PBKDF2 and PBES2 impls in SunJCE provider

### PR DESCRIPTION
**Pull request by valeriepeng**
This RFE enhances existing PBE algorithms with the "SHA512/224" and "SHA512/256" support. 
Current transformation parsing in javax.crypto.Cipher class is re-written to handle the additional "/" in the "SHA512/224" and "SHA512/256" algorithm names. Existing tests are updated with the additional new algorithms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8297538](https://bugs.openjdk.org/browse/JDK-8297538) to be approved

### Issues
 * [JDK-8288050](https://bugs.openjdk.org/browse/JDK-8288050): Add support of SHA-512/224 and SHA-512/256 to the PBKDF2 and PBES2 impls in SunJCE provider
 * [JDK-8297538](https://bugs.openjdk.org/browse/JDK-8297538): Add support of SHA-512/224 and SHA-512/256 to the PBKDF2 and PBES2 impls in SunJCE provider (**CSR**)

### Reviewers
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to [2bf57b16](https://git.openjdk.org/jdk/pull/11339/files/2bf57b163fc5456fd39bbf567cf789c3c6b61526)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11339/head:pull/11339` \
`$ git checkout pull/11339`

Update a local copy of the PR: \
`$ git checkout pull/11339` \
`$ git pull https://git.openjdk.org/jdk pull/11339/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11339`

View PR using the GUI difftool: \
`$ git pr show -t 11339`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11339.diff">https://git.openjdk.org/jdk/pull/11339.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/com/sun/crypto/provider/PBEKeyFactory.java [hunk 1]
```diff
@@ -164,6 +168,20 @@ public PBEWithHmacSHA512AndAES_128()  {
...
+    public static final class PBEWithHmacSHA512_224AndAES_128
+            extends PBEKeyFactory {
+        public PBEWithHmacSHA512_224AndAES_128()  {
+            super("PBEWithHmacSHA512/224AndAES_128");
+        }
+    }
+
+    public static final class PBEWithHmacSHA512_256AndAES_128
+            extends PBEKeyFactory {
+        public PBEWithHmacSHA512_256AndAES_128()  {
```
Comment (turbanoff) [inline src/java.base/share/classes/com/sun/crypto/provider/PBEKeyFactory.java:180]: Nit:
```suggestion
        public PBEWithHmacSHA512_256AndAES_128() {
```

Comment (valeriepeng) [reply src/java.base/share/classes/com/sun/crypto/provider/PBEKeyFactory.java:180]: Sure, will fix.

File: src/java.base/share/classes/javax/crypto/Cipher.java [hunk 2]
```diff
@@ -320,27 +322,30 @@ private static String[] tokenizeTransformation(String transformation)
...
          */
-        String[] parts = new String[3];
-        int count = 0;
-        StringTokenizer parser = new StringTokenizer(transformation, "/");
-        try {
-            while (parser.hasMoreTokens() && count < 3) {
-                parts[count++] = parser.nextToken().trim();
-            }
-            if (count == 0 || count == 2) {
+        String[] parts = { "", "", "" };
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply src/java.base/share/classes/javax/crypto/Cipher.java:325]: Yes, good catch. I added the additional algorithm to the existing PBES2Test and also updated the indexOf(...) call with additional toUpperCase(...) call.

**Comment by valeriepeng** [comment]
> I've added myself as a reviewer to the CSR. You might want to create a release note and a subtask to update the provider doc. Thanks.

Sure will, thanks for the reminder~

**Comment by ecki** [comment]
BTW while you touch the algorithm documentation, it Liste pkcs#5 which does work with octet strings, but the java api works with char passwords. It should describe that it uses the ucs2 char->bytes method not utf8 (pkcs8 vs pkcs12 if I recommend correctly). 

(Or even better allow both as currently i would still need to use BC for a byte representation in better defined utf8)

**Comment by valeriepeng** [comment]
> BTW while you touch the algorithm documentation, it Liste pkcs#5 which does work with octet strings, but the java api works with char passwords. It should describe that it uses the ucs2 char->bytes method not utf8 (pkcs8 vs pkcs12 if I recommend correctly).
> 
> (Or even better allow both as currently i would still need to use BC for a byte representation in better defined utf8)

Allow both would need a separate RFE. I will ask the doc team to document that only ASCII is supported for SunJCE provider for now. Thanks for the comment~

**Comment by ecki** [comment]
I checked again, it is in the Javadoc, but a bit hidden since it’s not listed near the algorithm name (in the JCE algorithm names doc).

the PBEKey limits it to single byte ASCII only: https://github.com/openjdk/jdk/blob/6dc4d891c3ad043405c65e0e0eeef28e9e5a2156/src/java.base/share/classes/com/sun/crypto/provider/PBEKey.java#L69 (which would be compat with utf8), that is indeed a good info to document that restriction as the Javadoc only says “takes low byte”.

https://docs.oracle.com/en/java/javase/17/docs/api/java.base/javax/crypto/spec/PBEKeySpec.html

it also mentions that the pkcs12 factory uses ucs2 (well “2 bytes”). But still would be nice to have that info in the algorithm spec doc as well.

Regarding a pkcs8 mode with utf8, what’s the easiest to propose a rfe for a non-commited?

**Comment by ecki** [comment]
Another question, the cipher/keyfactlry PBEWithHmacSHA512/224AndAES_256 algorithm looks a bit strange, that would require two rounds of the iterations - I mean it might not be much work to implement it anyway, but I wonder if it is a good combination?

On the other hand PBEWithHmacSHA512/224AndAES_192 might be a good combo. (Or generally the AES192 variants are missing?)

**Pull request closed without merging**